### PR TITLE
UCT/CUDA_COPY: Dummy registration of pinned/managed memory

### PR DESCRIPTION
**Pull request by Akshay-Venkatesh**
## What
Dummy registration for pinned/managed memory in cuda_copy.

## Why ?
Even though cuMemcpy operations don't need the registration, for cuda_copy put/get operations over pinned/managed memory we need to call uct_md_mem_reg in UCP flows. This PR populates memory handle with dummy value for pinned/managed memory.

### CODE REVIEW

**Comment by Akshay-Venkatesh** [comment]
@bureddy can you please provide a review?

File: src/uct/cuda/cuda_copy/cuda_copy_md.c [hunk 1]
```diff
@@ -85,9 +87,13 @@ UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_copy_mem_reg,
...
-    if ((result == CUDA_SUCCESS) && (memType == CU_MEMORYTYPE_HOST)) {
-        /* memory is allocated with cudaMallocHost which is already registered */
-        *memh_p = NULL;
+    if ((result == CUDA_SUCCESS) && ((memType == CU_MEMORYTYPE_HOST)    ||
+                                     (memType == CU_MEMORYTYPE_UNIFIED) ||
+                                     (memType == CU_MEMORYTYPE_DEVICE))) {
+        /* only host memory not allocated by cuda needs to be registered */
+        /* using deadbeef as VA to avoid gtest error */
+        UCS_STATIC_ASSERT((uint64_t)0xdeadbeef != (uint64_t)UCT_MEM_HANDLE_NULL);
+        *memh_p = (void *)0xdeadbeef;
```
Comment (bureddy) [inline src/uct/cuda/cuda_copy/cuda_copy_md.c:96]: why can't you directly assing UCT_MEM_HANDLE_NULL ?

Comment (Akshay-Venkatesh) [reply src/uct/cuda/cuda_copy/cuda_copy_md.c:96]: I saw some gtests also fail if mem_reg return handle_null

**Comment by brminich** [comment]
ok to test

**Comment by Akshay-Venkatesh** [comment]
@bureddy do you have other comments/feedback for the PR?

**Review by bureddy — APPROVED** [review]

**Comment by Akshay-Venkatesh** [comment]
@yosefe can you please review this PR?

**Comment by yosefe** [comment]
@Akshay-Venkatesh which tests fail or gets bad performance without this patch?
asking to make sure we have coverage in CI

**Comment by Akshay-Venkatesh** [comment]
> 
> 
> @Akshay-Venkatesh which tests fail or gets bad performance without this patch?
> asking to make sure we have coverage in CI

@yosefe Current tests don't fail or have bad performance because device-to-device transfers over ep_zcopy isn't stressed for UCP paths today AFAIK (there maybe at least one exception to this which is detailed below). This is needed when we have rendezvous pipeline over device buffers and we need to register user buffer before copying over to device bounce buffer. 

@bureddy what happens for d2d send-to-self path? Without https://github.com/openucx/ucx/pull/6115 cuda_copy shouldn't be used for send to self (if we do we're relying on undefined behaviour of cudaMemcpyAsync operations working when the wrong direction enum is used with it).

**Comment by yosefe** [comment]
@Akshay-Venkatesh do we have some UCT-level test for d2d put/get zcopy?

**Comment by bureddy** [comment]
@Akshay-Venkatesh  why do we need to register user buffer before copying over to device bounce buffer.  for cuda-copy we dont need it right?

**Comment by Akshay-Venkatesh** [comment]
> 
> 
> @Akshay-Venkatesh do we have some UCT-level test for d2d put/get zcopy?

Officially, only changes in https://github.com/openucx/ucx/pull/6115 would permit uct cuda-copy to realize d2d transfers. This is why I'd like to know how UCP realizes send-to-self when both src and dst is on GPU.

I'm not sure if there are generic UCT-level tests for d2d. 
1. I see a skip here https://github.com/Akshay-Venkatesh/ucx/blob/topic/cuda-copy-reg-non-host/test/gtest/uct/test_p2p_mix.cc#L169
2. I see d2d being checked for cuda-ipc case https://github.com/Akshay-Venkatesh/ucx/blob/topic/cuda-copy-reg-non-host/test/gtest/uct/test_p2p_rma.cc#L75 but this may not check intra-process d2d
3. Also see a skip here https://github.com/Akshay-Venkatesh/ucx/blob/topic/cuda-copy-reg-non-host/test/gtest/uct/test_zcopy_comp.cc#L44

**Comment by bureddy** [comment]
@Akshay-Venkatesh IMHO, #6115 is much more than permit uct cuda-copy to realize d2d transfers as it is trying to fix false cuda stream sharing.  d2d can be simply realized with cudaMemcpyDefault.  in fact, that is also not needed. cuda driver seems to ignore this parameter completely. in my PoC code, d2d worked w/o any changes to cuda_copy TL.

**Comment by Akshay-Venkatesh** [comment]
> 
> 
> @Akshay-Venkatesh why do we need to register user buffer before copying over to device bounce buffer. for cuda-copy we dont need it right?

cuMemcpyAsync doesn't and that's why we have a dummy registration in this PR. Doesn't UCP require registration of bounce buffers before copying over user buffers to bounce buffers? (similar to how we register host bounce buffers for 3-stage pipeline today. I've edited https://github.com/openucx/ucx/pull/6375#issuecomment-783559268 to correct that bounce buffers need registration and not the user buffers.)

**Comment by Akshay-Venkatesh** [comment]
> 
> 
> @Akshay-Venkatesh IMHO, #6115 is much more than permit uct cuda-copy to realize d2d transfers as it is trying to fix false cuda stream sharing. d2d can be simply realized with cudaMemcpyDefault. in fact, that is also not needed. cuda driver seems to ignore this parameter completely. in my PoC code, d2d worked w/o any changes to cuda_copy TL.

Agree that it works but as I've mentioned, this is not defined behaviour and it may lead issues going forward if this behaviour is fixed in a future revision. At minimum we need to change to MemcpyDefault enum. In #6115 I'm generalizing the current state of cuda_copy (which is also trying to avoid unnecessary serialization) to move D/H/M and not necessarily adding changes to fix false serialization as such.

**Comment by bureddy** [comment]
> > @Akshay-Venkatesh why do we need to register user buffer before copying over to device bounce buffer. for cuda-copy we dont need it right?
> 
> cuMemcpyAsync doesn't and that's why we have a dummy registration in this PR. Doesn't UCP require registration of bounce buffers before copying over user buffers to bounce buffers? (similar to how we register host bounce buffers for 3-stage pipeline today. I've edited [#6375 (comment)](https://github.com/openucx/ucx/pull/6375#issuecomment-783559268) to correct that bounce buffers need registration and not the user buffers.)

we dont need it today.  we register host bounce buffers just to pin the memory. but there is no rkey.
we pass NULL ( https://github.com/openucx/ucx/blob/master/src/ucp/rndv/rndv.c#L1758)

**Comment by Akshay-Venkatesh** [comment]
> 
> 
> > @Akshay-Venkatesh do we have some UCT-level test for d2d put/get zcopy?
> 
> Officially, only changes in #6115 would permit uct cuda-copy to realize d2d transfers. This is why I'd like to know how UCP realizes send-to-self when both src and dst is on GPU.
> 
> I'm not sure if there are generic UCT-level tests for d2d.
> 
>     1. I see a skip here https://github.com/Akshay-Venkatesh/ucx/blob/topic/cuda-copy-reg-non-host/test/gtest/uct/test_p2p_mix.cc#L169
> 
>     2. I see d2d being checked for cuda-ipc case https://github.com/Akshay-Venkatesh/ucx/blob/topic/cuda-copy-reg-non-host/test/gtest/uct/test_p2p_rma.cc#L75 but this may not check intra-process d2d
> 
>     3. Also see a skip here https://github.com/Akshay-Venkatesh/ucx/blob/topic/cuda-copy-reg-non-host/test/gtest/uct/test_zcopy_comp.cc#L44

@bureddy can you confirm if d2d transfers are tested in UCT-level put/get zcopy gtests to answer @yosefe 's question?
Also, what is the UCP-level set of operations to realize send-to-self? Is there registration and put/get zcopy involved? I know in the past that with OpenMPI, datatype component was somehow being used in send-to-self d2d flow but I don't see that on ucx master anymore.

**Comment by Akshay-Venkatesh** [comment]
> 
> 
> > > @Akshay-Venkatesh why do we need to register user buffer before copying over to device bounce buffer. for cuda-copy we dont need it right?
> > 
> > 
> > cuMemcpyAsync doesn't and that's why we have a dummy registration in this PR. Doesn't UCP require registration of bounce buffers before copying over user buffers to bounce buffers? (similar to how we register host bounce buffers for 3-stage pipeline today. I've edited [#6375 (comment)](https://github.com/openucx/ucx/pull/6375#issuecomment-783559268) to correct that bounce buffers need registration and not the user buffers.)
> 
> we dont need it today. we register host bounce buffers just to pin the memory. but there is no key.
> we pass NULL ( https://github.com/openucx/ucx/blob/master/src/ucp/rndv/rndv.c#L1758)

Going forward, we still need to pin if host buffers are used. Seems like registration is the only UCT level API that allows that. If there's another way to do that then I agree that we don't need to try and do dummy registration for cuda allocated memory.

**Comment by bureddy** [comment]
> > > > @Akshay-Venkatesh why do we need to register user buffer before copying over to device bounce buffer. for cuda-copy we dont need it right?
> > > 
> > > 
> > > cuMemcpyAsync doesn't and that's why we have a dummy registration in this PR. Doesn't UCP require registration of bounce buffers before copying over user buffers to bounce buffers? (similar to how we register host bounce buffers for 3-stage pipeline today. I've edited [#6375 (comment)](https://github.com/openucx/ucx/pull/6375#issuecomment-783559268) to correct that bounce buffers need registration and not the user buffers.)
> > 
> > 
> > we dont need it today. we register host bounce buffers just to pin the memory. but there is no key.
> > we pass NULL ( https://github.com/openucx/ucx/blob/master/src/ucp/rndv/rndv.c#L1758)
> 
> Going forward, we still need to pin if host buffers are used. Seems like registration is the only UCT level API that allows that. If there's another way to do that then I agree that we don't need to try and do dummy registration for cuda allocated memory.



> d2d tr

No. we have only UCP tests for mem type. 
on self endpoint,  ucp tag/secv uses cuda_copy TL (put/get_short and 3-stage pipelining).

**Comment by Akshay-Venkatesh** [comment]
>  on self endpoint, ucp tag/secv uses cuda_copy TL (put/get_short and 3-stage pipelining).

I don't see this behaviour (I see a single cuMemcpy for large transfers) and we wouldn't want that inefficient path to be used. Send-to-self should be done in one single copy without using any bounce buffers (unless this is forbidden for scenarios like lack of peer-accessibility or peer-accessiblility is [disabled](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__PEER__ACCESS.html#group__CUDA__PEER__ACCESS_1g5b4b6936ea868d4954ce4d841a3b4810) between device contexts when we support multiple device contexts per thread). 


@yosefe @bureddy regardless of the end use of uct_md_mem_reg through cuda_copy, doing a dummy registration for memory not allocated by cuda as per this PR should result in correct behaviour as far as what UCT API expects. I've confirmed this through gtests as well. That being the case, is it ok to discuss testing d2d UCT-level tests outside this PR and discuss that in #6115 (or some PR that correctly enables cuda_copy to move d2d)? If yes, I'd like to address any other outstanding issues in this PR if not already.

**Comment by yosefe** [comment]
> @yosefe @bureddy regardless of the end use of uct_md_mem_reg through cuda_copy, doing a dummy registration for memory not allocated by cuda as per this PR should result in correct behaviour as far as what UCT API expects. I've confirmed this through gtests as well. That being the case, is it ok to discuss testing d2d UCT-level tests outside this PR and discuss that in #6115 (or some PR that correctly enables cuda_copy to move d2d)? If yes, I'd like to address any other outstanding issues in this PR if not already.

Can we have at least UCT level tests for this PR?
Otherwise, how do we know it's working if this flow isn't being exercised in any test?

**Comment by yosefe** [comment]
Alternatively, this PR could be merged into $6115

**Comment by Akshay-Venkatesh** [comment]
> Can we have at least UCT level tests for this PR?
> Otherwise, how do we know it's working if this flow isn't being exercised in any test?

@yosefe uct_md_mem_query calls test/gtest/test_md.cc does test this PR with tests like (test_md, reg, ...) and (test_md, reg_perf, ...). This should suffice right?

**Comment by yosefe** [comment]
> @yosefe uct_md_mem_query calls test/gtest/test_md.cc does test this PR with tests like (test_md, reg, ...) and (test_md, reg_perf, ...). This should suffice right?

If it's actually testing cuda MD registering device memory, then yes

**Comment by Akshay-Venkatesh** [comment]
> 
> 
> > @yosefe uct_md_mem_query calls test/gtest/test_md.cc does test this PR with tests like (test_md, reg, ...) and (test_md, reg_perf, ...). This should suffice right?
> 
> If it's actually testing cuda MD registering device memory, then yes

It is testing cuda MD. Sorry I wasn't clear when I referred to that here: 

> @yosefe @bureddy regardless of the end use of uct_md_mem_reg through cuda_copy, doing a dummy registration for memory not allocated by cuda as per this PR should result in correct behaviour as far as what UCT API expects. I've confirmed this through gtests as well.

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@bureddy are you ok with this PR?

**Comment by bureddy** [comment]
👍

**Pull request merged**